### ISSUE TITLE
useLayoutEffect does nothing on the server

### ISSUE THREAD

**Issue by subashcs** [styled-components/styled-components#3649]
useLayoutEffect does nothing on the server warning is raised  on the console when using styled components with Next JS. 

## Environment
Next js project created with Next CLI and using styled components for styling.

```
{
  "dependencies": {
    "@ant-design/icons": "^4.7.0",
    "antd": "^4.17.4",
    "axios": "^0.24.0",
    "i18next": "^21.6.3",
    "next": "12.0.7",
    "react": "17.0.1",
    "react-dom": "17.0.1",
    "react-i18next": "^11.15.1",
    "react-redux": "^7.2.6",
    "redux": "^4.1.2",
    "redux-saga": "^1.1.3",
    "styled-components": "^5.3.3",
    "styled-normalize": "^8.0.7"
  },
  "devDependencies": {
    "@babel/core": "^7.16.5",
    "@commitlint/cli": "^15.0.0",
    "@commitlint/config-conventional": "^15.0.0",
    "@next/bundle-analyzer": "^12.0.7",
    "@storybook/addon-actions": "^6.4.9",
    "@storybook/addon-essentials": "^6.4.9",
    "@storybook/addon-links": "^6.4.9",
    "@storybook/addon-storyshots": "^6.4.9",
    "@storybook/react": "^6.4.9",
    "@types/enzyme": "^3.10.10",
    "@types/jest": "^27.0.3",
    "@types/node": "^17.0.1",
    "@types/react": "^17.0.37",
    "@types/react-dom": "^17.0.11",
    "@types/react-redux": "^7.1.20",
    "@types/redux-saga": "^0.10.5",
    "@types/styled-components": "^5.1.18",
    "@typescript-eslint/eslint-plugin": "^5.7.0",
    "@typescript-eslint/parser": "^5.7.0",
    "@wojtekmaj/enzyme-adapter-react-17": "^0.6.6",
    "babel-jest": "^27.4.5",
    "babel-loader": "^8.2.3",
    "babel-plugin-import": "^1.13.3",
    "babel-plugin-styled-components": "^2.0.2",
    "cross-env": "^7.0.3",
    "cz-conventional-changelog": "^3.3.0",
    "enzyme": "^3.11.0",
    "eslint": "^8.5.0",
    "eslint-config-airbnb": "^19.0.2",
    "eslint-config-prettier": "^8.3.0",
    "eslint-import-resolver-alias": "^1.1.2",
    "eslint-import-resolver-typescript": "^2.5.0",
    "eslint-plugin-import": "^2.25.3",
    "eslint-plugin-jest": "^25.3.0",
    "eslint-plugin-jsx-a11y": "^6.5.1",
    "eslint-plugin-react": "^7.27.1",
    "eslint-plugin-react-hooks": "^4.3.0",
    "eslint-plugin-unicorn": "^39.0.0",
    "husky": "^7.0.4",
    "identity-obj-proxy": "^3.0.0",
    "jest": "^27.4.5",
    "jest-specific-snapshot": "^5.0.0",
    "jest-styled-components": "^7.0.8",
    "lint-staged": "^12.1.3",
    "next-redux-wrapper": "^7.0.5",
    "prettier": "^2.5.1",
    "redux-devtools-extension": "^2.13.9",
    "standard-version": "^9.3.2",
    "storybook-addon-next-router": "^3.1.0",
    "stylelint": "^14.1.0",
    "stylelint-config-prettier": "^9.0.3",
    "stylelint-config-standard": "^24.0.0",
    "stylelint-config-styled-components": "^0.1.1",
    "stylelint-order": "^5.0.0",
    "stylelint-processor-styled-components": "^1.10.0",
    "typescript": "^4.5.4"
  }
}

```

## Reproduction
The bug can be reproduce when rendering styled components on the Server Side.

## Steps to reproduce
- Create a new next.js project. 
- Create a layout file that renders on the server.
- Import styled components on the layout file and use them.

## Expected Behavior
useLayoutEffect should not throw error while rendering styled components on the server.

## Actual Behavior
Getting this error on the terminal.

```
Warning: useLayoutEffect does nothing on the server, because its effect cannot be encoded into the server renderer's output format. This will lead to a mismatch between the initial, non-hydrated UI and the intended UI. To avoid this, useLayoutEffect should only be used in components that render exclusively on the client. See https://reactjs.org/link/uselayouteffect-ssr for common fixes.
    at Overflow (/home/subashcs/Documents/projects/nextjs-ts-antd-redux-storybook-starter/node_modules/rc-overflow/lib/Overflow.js:42:32)
    at InheritableContextProvider (/home/subashcs/Documents/projects/nextjs-ts-antd-redux-storybook-starter/node_modules/rc-menu/lib/context/MenuContext.js:40:23)
    at Menu (/home/subashcs/Documents/projects/nextjs-ts-antd-redux-storybook-starter/node_modules/rc-menu/lib/Menu.js:78:32)
    at InternalMenu (/home/subashcs/Documents/projects/nextjs-ts-antd-redux-storybook-starter/node_modules/antd/lib/menu/index.js:77:37)
    at Menu (/home/subashcs/Documents/projects/nextjs-ts-antd-redux-storybook-starter/node_modules/antd/lib/menu/index.js:174:37)
    at NavMenu (webpack-internal:///./src/components/Header/NavMenu.tsx:22:3)
    at header
    at Header__HeaderWrapper (/home/subashcs/Documents/projects/nextjs-ts-antd-redux-storybook-starter/node_modules/styled-components/dist/styled-components.cjs.js:1:19220)
    at Header (webpack-internal:///./src/components/Header/index.tsx:23:3)
    at Layout (webpack-internal:///./src/components/Layout/index.tsx:25:3)
    at WrappedApp (webpack-internal:///./pages/_app.tsx:35:3)
    at Provider (/home/subashcs/Documents/projects/nextjs-ts-antd-redux-storybook-starter/node_modules/react-redux/lib/components/Provider.js:21:20)
    at withRedux(WrappedApp) (/home/subashcs/Documents/projects/nextjs-ts-antd-redux-storybook-starter/node_modules/next-redux-wrapper/lib/index.js:184:35)
    at InnerApp
    at StyleRegistry (/home/subashcs/Documents/projects/nextjs-ts-antd-redux-storybook-starter/node_modules/styled-jsx/dist/stylesheet-registry.js:231:34)
    at AppContainer (/home/subashcs/Documents/projects/nextjs-ts-antd-redux-storybook-starter/node_modules/next/dist/server/render.js:340:29)
    at AppContainerWithIsomorphicFiberStructure (/home/subashcs/Documents/projects/nextjs-ts-antd-redux-storybook-starter/node_modules/next/dist/server/render.js:370:57)
    at div
    at Body (/home/subashcs/Documents/projects/nextjs-ts-antd-redux-storybook-starter/node_modules/next/dist/server/render.js:638:21)
```
## Possible Solution
Wherever the useLayoutEffect hook is being used replace it with useEffect hook if its being rendered on the server. May be adding a check like if window is defiined then use useLayoutEffect otherwise useEffect.

**Comment by subashcs** [comment]
The error had occured because I had created the styled component inside the React component. 
Closing the issue as styled-component library also suggests to create components outside react components. This worked to mitigate the useLayoutEffect error thrown by styled-components. However the issue also existed on antd components.
